pkg/credentials/buffer/ring_buffer_test.go

```diff
@@ -0,0 +1,343 @@
+		t.Run(name, func(t1 *testing.T) {
+			buf := buffer.NewRingBuffer(tc.capacity)
+
+			buf.AddMulti(tc.bytesToAdd)
+
+			isFull := buf.IsFull()
+
+			if isFull != tc.expectedFull {
+				t1.Errorf("expected buffer full %t, but was not", tc.expectedFull)
+			}
+		})
+	}
+}
+
+func TestIsEmpty(t *testing.T) {
+	type isEmptyTest struct {
+		capacity      int
+		bytesToAdd    []byte
+		expectedEmpty bool
+	}
+	testCases := []isEmptyTest{
+		{0, []byte{}, true},
+		{1, []byte{}, true},
+		{2, []byte{}, true},
+		{3, []byte{}, true},
+		{1, []byte{1}, false},
+		{2, []byte{1}, false},
+		{2, []byte{1, 2}, false},
+		{3, []byte{1, 2}, false},
+		{3, []byte{1, 2, 3}, false},
+	}
+
+	for _, tc := range testCases {
+		name := fmt.Sprintf("cap_%d_empty_%t_withbytes_%v", tc.capacity, tc.expectedEmpty, tc.bytesToAdd)
+		t.Run(name, func(t1 *testing.T) {
+			buf := buffer.NewRingBuffer(tc.capacity)
+
+			buf.AddMulti(tc.bytesToAdd)
+
+			isEmpty := buf.IsEmpty()
+
+			if isEmpty != tc.expectedEmpty {
+				t1.Errorf("expected buffer empty %t, but was not", tc.expectedEmpty)
+			}
+		})
+	}
+}
+
+func TestNotIsEmptyAfterPoppingPartially(t *testing.T) {
+	buf := buffer.NewRingBuffer(3)
+
+	buf.Add(1)
+	buf.Add(2)
+	buf.Add(3)
+
+	_, err := buf.Pop(2)
+	if err != nil {
+		t.Fatal(err)
+	}
+	empty := buf.IsEmpty()
+	if empty {
+		t.Error("Expected buffer to be not empty")
+	}
+}
+
+func TestIsEmptyAfterPoppingAllBytes(t *testing.T) {
+	buf := buffer.NewRingBuffer(2)
+
+	buf.Add(1)
+	buf.Add(2)
+
+	_, err := buf.PopFirst()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = buf.PopFirst()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	isEmpty := buf.IsEmpty()
+	if !isEmpty {
+		t.Error("expected buffer to be empty at end")
+	}
+}
+
+func TestGet(t *testing.T) {
+	type getTest struct {
+		capacity        int
+		bytesToAdd      []byte
+		indexToRetrieve int
+		expectedByte    byte
+		expectedError   bool
+	}
+	testCases := []getTest{
+		{0, []byte{}, -1, 0, true},
+		{0, []byte{}, 0, 0, true},
+		{1, []byte{}, 1, 0, true},
+		{1, []byte{}, -1, 0, true},
+		{3, []byte{1, 2}, 2, 0, true},
+		{3, []byte{1, 2, 3}, 0, 1, false},
+		{3, []byte{1, 2, 3}, 1, 2, false},
+		{3, []byte{1, 2, 3}, 2, 3, false},
+		{3, []byte{1, 2, 3, 4}, 2, 4, false},    // overflow buffer
+		{3, []byte{1, 2, 3, 4, 5}, 2, 5, false}, // overflow buffer
+	}
+
+	for _, tc := range testCases {
+		name := fmt.Sprintf("cap_%d_index_%d_byte_%b_err_%t_withbytes_%v", tc.capacity,
+			tc.indexToRetrieve, tc.expectedByte, tc.expectedError, tc.bytesToAdd)
+		t.Run(name, func(t1 *testing.T) {
+			buf := buffer.NewRingBuffer(tc.capacity)
+
+			buf.AddMulti(tc.bytesToAdd)
+
+			b, err := buf.Get(tc.indexToRetrieve)
+			if tc.expectedError && err == nil {
+				t1.Error("expected error but got nil")
+			} else if !tc.expectedError && err != nil {
+				t1.Error(err)
+			} else if tc.expectedByte != b {
+				t1.Errorf("expected byte %b but got %b", tc.expectedByte, b)
+			}
+
+			if !tc.expectedError {
```

Comment:
Suggest adding a comment that this block is here to ensure that `get` doesn't modify the buffer